src/Session.cc

```diff
@@ -560,9 +560,19 @@ const AddressSpace::Mapping Session::recreate_shared_mmap(
   uint32_t flags = m.flags;
   size_t size = m.map.size();
   void* preserved_data = preserve == PRESERVE_CONTENTS ? m.local_addr : nullptr;
-  if (preserved_data) {
-    remote.task()->vm()->detach_local_mapping(m.map.start());
-  }
+
+  // Note that Mapping `m` may correspond to a Mapping from a different Task than
+  // the `maybe_detach_mapping`. See replay_syscall.cc prepare_clone() for an example.
+  auto maybe_detach_mapping = remote.task()->vm()->mapping_of(m.map.start());
+
+  // Sanity check
+  ASSERT(remote.task(), size == maybe_detach_mapping.map.size());
+  ASSERT(remote.task(), m.map.start() == maybe_detach_mapping.map.start());
+
+  // Note that both preserved_data == maybe_local_unmap or preserved_data <>
+  // maybe_local_unmap cases are possible
+  void* maybe_local_unmap = remote.task()->vm()->detach_local_mapping(m.map.start());
```

Comment:
The comment and namng here are a bit vague. I think better to call this `remote_task_local_mapping` and just delete the comment.

Reply:
Done